include/deal.II/base/symbolic_function.templates.h

```diff
@@ -0,0 +1,243 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_symbolic_function_templates_h
+#define dealii_symbolic_function_templates_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/symbolic_function.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Functions
+{
+#ifdef DEAL_II_WITH_SYMENGINE
+
+  namespace
+  {
+    inline std::vector<Differentiation::SD::Expression>
+    to_expressions(const std::vector<std::string> &strings)
+    {
+      std::vector<Differentiation::SD::Expression> ret(strings.size());
+      for (unsigned int i = 0; i < strings.size(); ++i)
+        ret[i] = Differentiation::SD::Expression(strings[i], true);
+      return ret;
+    }
+  } // namespace
+
+
+
+  template <int dim, typename RangeNumberType>
+  SymbolicFunction<dim, RangeNumberType>::SymbolicFunction(
+    const std::vector<Differentiation::SD::Expression> &   functions,
+    const Tensor<1, dim, Differentiation::SD::Expression> &argument,
+    const Differentiation::SD::Expression &                time,
+    const Differentiation::SD::types::substitution_map &user_substitution_map)
+    : Function<dim, RangeNumberType>(functions.size(), 0.0)
+    , user_function(functions)
+    , user_substitution_map(user_substitution_map)
+    , coordinate_symbols(argument)
+    , time_symbol(time)
+  {
+    update_derivatives();
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  SymbolicFunction<dim, RangeNumberType>::SymbolicFunction(
+    const std::string &expressions)
+    : Function<dim, RangeNumberType>(
+        Utilities::split_string_list(expressions, ';').size(),
+        0.0)
+    , user_function(
+        to_expressions(Utilities::split_string_list(expressions, ';')))
+    , coordinate_symbols(get_default_coordinate_symbols())
+    , time_symbol(Differentiation::SD::make_symbol("t"))
+  {
+    update_derivatives();
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  void
+  SymbolicFunction<dim, RangeNumberType>::update_derivatives()
+  {
+    function.resize(user_function.size());
+    function_gradient.resize(user_function.size());
+    function_hessian.resize(user_function.size());
+    function_laplacian.resize(user_function.size());
+
+    for (unsigned int i = 0; i < user_function.size(); ++i)
+      {
+        function[i] = user_function[i].substitute(user_substitution_map);
```

Comment:
I really don't think that this is necessary. I don't see why the `user_substitution_map` cannot be added to `create_substitution_map_for_point()`. If the user symbols evaluate to constants, or some polynomial function of another user symbol, then we're just doing all of the costly work of differentiation for no reason. If the user symbols depend on the time or evaluation coordinates, then we have a mechanism for them to state that dependency by declaring the symbol a symbolic function. We don't do any partial differentiation here, so there's certainly not an intermediate step that requires substitution before computing a total derivative.

Could you maybe describe a use case or give a short piece of pseudo-code so that I can better understand why your approach would be necessary/preferable to performing the differentiation only once per instance that a `SymbolicFunction` is created?

Reply:
My reasoning is the following: I'd like function/gradient/hessian evaluation to be as quick as possible.  For that evaluation to be quick, I don't want to have to compute derivatives *at every evaluation point*. Think of this case:

~~~
SymbolicFunction<2> my_function("r^3+g(theta)");
auto my_map = make_substitution_map("r", Expression("sqrt(x^2+y^2)", true));
my_function.update_user_substitution_map(my_map);
...
my_function.gradient(p);
~~~
with your approach, I'd have to compute the gradient on the fly, before substituting the value of the point p. We could compute the gradients and hessians only if required (this I can fix), but I still believe it is better *not* to compute them at every point evaluation. With your suggestion, we'd get a gradient computation for every evaluation...

Reply:
Ok, fair point. But I think that we can easily compromise on this. We have two member functions, something like `set_user_substitution_map()` that does what I want to do, and `perform_permanent_substitution()` that does what you want to do.  I'll explain why in the reply to https://github.com/dealii/dealii/pull/9171#discussion_r368151217.

Reply:
As an aside, we could make a follow up issue / PR to add some functions that convert between Cartesian/cylindrical/spherical coordinates to the `GridTools` or `Physics` namespace, and make some member function that does this transformation for one (assuming some preset variables for the radius, angle etc. that can also be returned in the same way that `get_default_coordinate_symbols()` returns `x,y,z`).

Reply:
As an aside, we could make a follow up issue / PR to add some functions that convert between Cartesian/cylindrical/spherical coordinates to the `GridTools` or `Physics` namespace, and make some member function that does this transformation for one (assuming some preset variables for the radius, angle etc. that can also be returned in the same way that `get_default_coordinate_symbols()` returns `x,y,z`).

Reply:
I think there is a misunderstanding on how things work now.

There are two objects. One is user_function, immutable, inserted at construction time, that contains all the symbols one wants. The other is function, which is the one that is actually evaluated. Since the only way to update the user_substitution_map is through the update_user_substitution_map, this function does not touch the user_function, but only the one that is used to evaluate the function, thus allowing you to perform other substitutions later on (by calling again update_user_substitution_map).
This puts together the advantages of both worlds. It does not change what the user inserted permanently, but only up to the next update_user_substitution_map. How this happen internally is now efficient, while with your suggestion it would be performed at every point evaluation.
Wasn't this what you suggested?